hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/impl/StateBase.java

```diff
@@ -0,0 +1,99 @@
+/*
+ * Copyright (C) 2022 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.hedera.node.app.state.impl;
+
+import com.hedera.node.app.spi.state.ReadableState;
+import com.hedera.node.app.spi.state.WritableState;
+import java.util.*;
+import javax.annotation.Nonnull;
+
+/**
+ * A base class for implementations of {@link ReadableState} and {@link WritableState}.
+ *
+ * @param <K> The key type
+ * @param <V> The value type
+ */
+abstract class StateBase<K, V> implements ReadableState<K, V> {
+    /** The state key, which cannot be null */
+    private final String stateKey;
+
+    /**
+     * A cache of all values read from this {@link ReadableState}. If the same value is read twice,
+     * rather than going to the underlying merkle data structures to read the data a second time, we
+     * simply return it from the cache. We also keep track of all keys read, which is critical for
+     * dealing with validating what we read during pre-handle with what may have changed before we
+     * got to handle transaction.
+     */
+    private final Map<K, V> readCache = new HashMap<>();
+
+    /**
+     * Create a new StateBase.
+     *
+     * @param stateKey The state key. Cannot be null.
+     */
+    StateBase(@Nonnull String stateKey) {
+        this.stateKey = Objects.requireNonNull(stateKey);
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Nonnull
+    public final String getStateKey() {
+        return stateKey;
+    }
+
+    /** {@inheritDoc} */
+    @Override
+    @Nonnull
+    public final Optional<V> get(@Nonnull K key) {
+        Objects.requireNonNull(key);
```

Comment:
Do we really need this check? If yes we should add it to contains too.

Reply:
As long as NPE gets thrown anyway, we don't need the line. I will check.